spec/components/middleware/request_tracker_spec.rb

```diff
@@ -386,6 +394,114 @@ def warn(*args)
+        x = 0
+        app = lambda do |_|
+          x += 1
+          [200, {}, ["OK"]]
+        end
+        env = env("REMOTE_ADDR" => "1.1.1.1")
+        middleware = Middleware::RequestTracker.new(app)
+        status, = middleware.call(env)
+        expect(status).to eq(200)
+        expect(x).to eq(1)
+
+        env = env("REMOTE_ADDR" => "1.1.1.1")
+        middleware = Middleware::RequestTracker.new(app)
+        status, headers, response = middleware.call(env)
+        expect(status).to eq(429)
+        expect(x).to eq(1)
+        expect(headers["Discourse-Rate-Limit-Error-Code"]).to eq("ip_60_secs_limit")
+        expect(response.first).to include("Error code: ip_60_secs_limit.")
+      end
+
+      it "is included when the assets-requests-per-10-seconds limit is reached" do
+        global_setting :max_asset_reqs_per_ip_per_10_seconds, 1
+        x = 0
+        app = lambda do |env|
+          x += 1
+          env["DISCOURSE_IS_ASSET_PATH"] = true
+          [200, {}, ["OK"]]
+        end
+        env = env("REMOTE_ADDR" => "1.1.1.1")
+        middleware = Middleware::RequestTracker.new(app)
+        status, = middleware.call(env)
+        expect(status).to eq(200)
+        expect(x).to eq(1)
+
+        env = env("REMOTE_ADDR" => "1.1.1.1")
+        middleware = Middleware::RequestTracker.new(app)
+        status, headers, response = middleware.call(env)
+        expect(status).to eq(429)
+        expect(x).to eq(1)
+        expect(headers["Discourse-Rate-Limit-Error-Code"]).to eq("ip_assets_10_secs_limit")
+        expect(response.first).to include("Error code: ip_assets_10_secs_limit.")
+      end
+    end
+
+    it "users with high enough trust level are not rate limited per ip" do
+      global_setting :max_reqs_per_ip_per_minute, 1
+      global_setting :skip_per_ip_rate_limit_trust_level, 3
+
+      envs = 3.times.map do |n|
+        user = Fabricate(:user, trust_level: 3)
+        token = UserAuthToken.generate!(user_id: user.id)
+        cookie = DiscourseAuthCookie.new(
+          token: token.unhashed_auth_token,
+          user_id: user.id,
+          trust_level: user.trust_level,
+          timestamp: Time.zone.now,
+          valid_for: 10.hours
+        ).to_text
+        env("HTTP_COOKIE" => "_t=#{cookie}", "REMOTE_ADDR" => "1.1.1.1")
+      end
+
+      x = 0
+      app = lambda do |env|
+        x += 1
+        [200, {}, ["OK"]]
+      end
+      envs.each do |env|
+        middleware = Middleware::RequestTracker.new(app)
+        status, = middleware.call(env)
+        expect(status).to eq(200)
+      end
+      expect(x).to eq(3)
+
+      envs.each do |env|
+        middleware = Middleware::RequestTracker.new(app)
+        status, headers, response = middleware.call(env)
+        expect(status).to eq(429)
+        expect(headers["Discourse-Rate-Limit-Error-Code"]).to eq("id_60_secs_limit")
+        expect(response.first).to include("Error code: id_60_secs_limit.")
+      end
+      expect(x).to eq(3)
```

Comment:
I'm not sure I follow here, Shouldn't each user be able to make as many requests as they want since they will skip the rate limit?

Reply:
No, the rate limits are not skipped entirely for trusted users. Right now we blindly apply rate limits per IP so if 3 users share the same IP and we allow 200 requests per IP per minute then the 3 users share the 200 requests. With this change, each user gets their own 200 requests.

Reply:
Ah got it. I re-read the setting and understand it better now. Basically if a certain trust level is met, the user will be using a per user rate limit rather than a per ip rate limit.

Reply:
Hmm something is still confusing for me. Even though they use a per user rate limit, that per user rate limit value is still tied to `max_reqs_per_ip_per_minute`?

Reply:
Yes. Right now `max_reqs_per_ip_per_minute` determines the max requests per IP per minute and the max requests per user per minute. Maybe we should add a new setting for per user?

Reply:
Yea a new setting probably makes sense because a rename will be painful. Not sure, maybe @SamSaffron have some thoughts on this?